apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PgNumeric.cs

```diff
@@ -431,15 +465,70 @@ public static PgNumeric FromDecimal(decimal value)
         /// <returns>The converted value.</returns>
         public decimal ToDecimal(LossOfPrecisionHandling lossOfPrecisionHandling)
         {
-            // FIXME: Awful performance!
-            string text = ToString();
-            // Handles overflow.
-            decimal dec = Decimal.Parse(text, CultureInfo.InvariantCulture);
-            if (lossOfPrecisionHandling == LossOfPrecisionHandling.Throw && this != dec)
+            if (_notANumber)
             {
-                throw new InvalidOperationException("Conversion would lose information");
+                throw new OverflowException($"NaN cannot be represented as {typeof(decimal).FullName}");
+            }
+
+            // We don't handle zero explicitly as we want to take into account trailing zeros here.
+
+            // Handles overflow
+            return decimal.Parse(ToString(DecimalPrecision), CultureInfo.InvariantCulture);
+
+            int DecimalPrecision(string scaledText, int integerPlaces, int decimalPlaces)
+            {
+                int decimalPrecision;
+                // If the number had no decimal places to begin with, then we need not to worry.
+                // If the number has 29 or more integer places we cannot include any decimal places;
+                // this may overflow when we convert to decimal but that's fine, let's not handle that here.
+                if (decimalPlaces == 0 || integerPlaces >= 29)
+                {
+                    decimalPrecision = 0;
+                }
+                // If the number has at most 28 significant digits we can represent it fully.
+                else if (integerPlaces + decimalPlaces <= 28)
+                {
+                    decimalPrecision = decimalPlaces;
+                }
+                // If there are no integer places we can represent up to 28 decimal places.
```

Comment:
Can we not represent 29 decimal places in some cases? or does the structure of decimal mean that 29 significant digits always includes one before the decimal point?

Reply:
This was a slight surprise for me as well (when I did BigQuery), but apparently "there's always one digit before the decimal point" as demonstrated by the following two snippets:

``` csharp
decimal d = decimal.Parse("0.12345678901234567890123456789");
Console.WriteLine(d);
```
writes (decimal parsing rounds instead of truncating but still proves the point):
```
0.1234567890123456789012345679
```

but:

```
decimal o = decimal.Parse("12345678901234567890123456789");
Console.WriteLine(o);
```
doesn't overflow, which is expected, but makes the above case surprising, instead it just prints the whole number:
```
12345678901234567890123456789
```

It really is unexpected as a decimal is supposed to be stored as a 96 bit integer + 32 bits of flags for sign and scale, so I really don't understand why that 0 before the decimal point is counting towards the amount of digits.

If I didn't know better I'd think there was a bug in `decimal.Parse` :) that's counting that zero as relevant.

Or I'm really missunderstanding something here, but the fact that it jumped at you as well makes me think I'm not.

Reply:
Remind me to have another look at this later in the week. (It needn't block merging.) It would be interesting to try doing this from the raw integers...

Reply:
Yep, I continued to tried things out, and if there's a bug (unlikely) is not in `decimal.Parse`:

```
decimal e = 0.12345678901234567890123456789m;
Console.WriteLine(e);
```

also prints the 28 significant digit rounded number:

```
0.1234567890123456789012345679
```

This is really surprising. I'll fidle with the bits more actually and see what I find.

Reply:
Just for posterity, there's nothing surprising going on here, although `0.12345678901234567890123456789` scaled, i.e. `12345678901234567890123456789` would really fit on the decimal 96bits for the scaled integer, the exponent would need to be `29` which is greater than the max possible exponent `28`.